zrml/swaps/src/root.rs

```diff
@@ -0,0 +1,308 @@
+//     works are solely in the form of machine-executable object code generated by
+//     a source language processor.
+//
+//     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//     IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//     FITNESS FOR A PARTICULAR PURPOSE, TITLE AND NON-INFRINGEMENT. IN NO EVENT
+//     SHALL THE COPYRIGHT HOLDERS OR ANYONE DISTRIBUTING THE SOFTWARE BE LIABLE
+//     FOR ANY DAMAGES OR OTHER LIABILITY, WHETHER IN CONTRACT, TORT OR OTHERWISE,
+//     ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+//     DEALINGS IN THE SOFTWARE.
+
+use sp_runtime::traits::AtLeast32BitUnsigned;
+
+/// Calculate the preimage of `value` under `f` using bisection.
+///
+/// This function always terminates (regardless of the properties of `f`), but `f` must be
+/// a monotonous function to guarantee that the result is correct. The algorithm _assumes_ that `f`
+/// is monotonous, but we cannot ensure this. Calling with a non-monotonous function is undefined
+/// behavior. We do some sanity checks, but in the end, the consumer is responsible for providing a
+/// function which satisfies the API.
+///
+/// We return `(preimage, iteration_count)`. If `[min, max]` contains no preimage, then
+/// `min` or `max` is returned, depending on which is closer to the preimage. If `[min, max]`
+/// contains a preimage, an approximation of a preimage is returned.
+///
+/// # Arguments
+///
+/// - `f`: The function
+/// - `value`: The value of the preimage
+/// - `min`: The minimum value of the preimage
+/// - `max`: The maximum value of the preimage
+/// - `max_iterations`: Break after this many iterations
+/// - `tol`: Break if the interval is smaller than this
+pub(crate) fn calc_preimage<T, F>(
+    f: F,
+    value: T,
+    mut min: T,
+    mut max: T,
+    max_iterations: usize,
+    tol: T,
+) -> Result<(T, usize), &'static str>
+where
+    T: AtLeast32BitUnsigned + Copy,
+    F: Fn(T) -> Result<T, &'static str>,
+{
+    if min >= max {
+        return Err("calc_preimage: Function domain has no volume");
+    }
+    let fmin = f(min)?;
+    let mut fmax = f(max)?;
+    if fmin == value {
+        return Ok((min, 0));
+    } else if fmax == value {
+        return Ok((max, 0));
+    }
+
+    if is_outside_of(value, fmin, fmax) {
+        if dist(fmax, value) < dist(fmin, value) {
+            return Ok((max, 0));
+        } else {
+            return Ok((min, 0));
+        }
+    }
+    // At this point we can assume that there exists a preimage!
+
+    // Defensively use `for` instead of `while` or `loop` to ensure that it breaks after
+    // `max_iterations`.
+    let mut mid = T::zero();
+    let mut iteration_count = 0;
+    for i in 1..=max_iterations {
+        iteration_count = i;
+        let size = max.checked_sub(&min).ok_or("Unexpected arithmetic underflow")?;
+        if size < tol {
+            break;
```

Comment:
I think you forgot to set the `mid` here for the case, that it breaks in the first iteration. 

`mid` should be `min or max`, but not `T::zero()`, right?

Reply:
Yeah, good catch. 064b2d321fd876d02e85bd0fca635fab9d6d6ddf

Reply:
If I could do it all over again, I'd let `calc_preimage` return the final `(min, max)` (like the boost implementation does) and let the consumer select which one they want to use.

Reply:
I serious considered doing this right now for about half an hour, but decided that I shouldn't do it, because that would mean rewriting part of the code you already checked, and also rewriting virtually all tests, because the approximations will be slightly different.

There's a slight advantage to be had from this, which is that you can decide whether or not you would rather go with a more conservative arbitrage amount (choose `min`) or risk overshooting the target (choose `max`). But that would only matter on pools with huge amount of liquidity.

Reply:
It's okay as it is. 🙃